README.md
# Helpful Tweaks

![](doc/assets/logo.png)

Mod for Minecraft to allow crafting and smelting vanilla items.

All options are fully configurable (enable/disable). Take a look into the config file: `helpfultweaks.cfg`.

**Current crafting and smelting recipes**

+ Slime ball
+ Web
+ String
+ Glowstone
+ Gunpowder
+ Nether Wart
+ Nether Quartz
+ Prismarine Crystal (x2)
+ Prismarine Shard (x2)
+ Charcoal
+ Flint
+ Leather

All recipes are available through [Just Enough Items](https://minecraft.curseforge.com/projects/just-enough-items-jei) (JEI).

---

## Crafting

### Slime Ball

![](doc/img/slimeball_recipe.jpg)

### Web

![](doc/img/web_recipe.jpg)

### String

![](doc/img/string_recipe.jpg)

### Glowstone

![](doc/img/glowstone_recipe.jpg)

### Gunpowder

![](doc/img/gunpowder_recipe.jpg)

### Nether Wart

![](doc/img/netherwart_recipe.jpg)

### Nether Quartz

![](doc/img/netherquartz_recipe.jpg)

### Prismarine Crystal (x2)

![](doc/img/prismarinecrystal_recipe.jpg)

### Prismarine Shard (x2)

![](doc/img/prismarineshard_recipe.jpg)

---

## Smelting

### Charcoal

![](doc/img/charcoal_recipe.jpg)

### Flint

![](doc/img/flint_recipe.jpg)

### Leather

![](doc/img/leather_recipe.jpg)

---

## License

This software is licensed under the MIT license.

---

## Feedback

Any suggestion is welcome!



CHANGELOG
v1.6.0  (March 2, 2017)

 - Added nether wart recipe.
 - Added nether quartz recipe.
 - Added prismarine crystal recipe (x2).
 - Added prismarine shard recipe (x2).



v1.5.0  (March 2, 2017)

 - Added glowstone recipe.
 - Added gunpowder recipe.



v1.4.1  (March 1, 2017)

 - Rebalancing web recipe.



v1.4.0  (March 1, 2017)

 - Added string to web recipe.
 - Added web to string recipe.



v1.3.0  (February 28, 2017)

 - Added rotted flesh to leather recipe.



v1.2.0  (February 27, 2017)

 - Added vanilla saplings to charcoal recipe.
 - Added enable/disable recipe functionality from config file
   (helpfultweaks.cfg).



v1.1.0  (February 27, 2017)

 - Added slime ball recipe.



v1.0.0  (February 3, 2017)

 - Added gravel to flint recipe.
 - First release.


src/main/java/com/icaroperseo/helpfultweaks/HelpfulTweaks.java
package com.icaroperseo.helpfultweaks;

/**
 * Created by icaroperseo on 1/03/17.
 */

import com.icaroperseo.helpfultweaks.common.proxy.CommonProxy;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.common.Mod.EventHandler;
import net.minecraftforge.fml.common.SidedProxy;
import net.minecraftforge.fml.common.event.FMLInitializationEvent;
import net.minecraftforge.fml.common.event.FMLPreInitializationEvent;

@Mod(modid = HelpfulTweaks.MODID, name = HelpfulTweaks.NAME, version = HelpfulTweaks.VERSION,
        acceptedMinecraftVersions = HelpfulTweaks.MCVERSION)
public class HelpfulTweaks {
    public static final String MODID = "helpfultweaks";
    public static final String NAME = "Helpful Tweaks";
    public static final String VERSION = "1.6.0";
    public static final String MCVERSION = "1.11.2";

    @Mod.Instance(HelpfulTweaks.MODID)
    public static HelpfulTweaks instance;

    @SidedProxy(clientSide="com.icaroperseo.helpfultweaks.client.core.proxy.ClientOnlyProxy",
            serverSide="com.icaroperseo.helpfultweaks.common.proxy.CommonProxy")
    public static CommonProxy proxy;

    @EventHandler
    public void preInit(FMLPreInitializationEvent event) {
        proxy.preInit();
    }

    @EventHandler
    public void init(FMLInitializationEvent event) {
        proxy.init();
    }
}


src/main/java/com/icaroperseo/helpfultweaks/client/crafting/recipe/HelpfulTweaksRecipes.java
package com.icaroperseo.helpfultweaks.client.crafting.recipe;

/**
 * Created by icaroperseo on 1/03/17.
 */

import com.icaroperseo.helpfultweaks.client.core.config.ConfigHandler;
import net.minecraft.init.Items;
import net.minecraft.init.Blocks;
import net.minecraft.item.ItemStack;
import net.minecraftforge.fml.common.registry.GameRegistry;

public class HelpfulTweaksRecipes {
    public static void addRecipes() {

        if (ConfigHandler.enableSlimeBallRecipe) {
            addSlimeBallRecipe();
        }

        if (ConfigHandler.enableCharcoalRecipe) {
            addCharcoalRecipe();
        }

        if (ConfigHandler.enableFlintRecipe) {
            addFlintRecipe();
        }

        if (ConfigHandler.enableLeatherRecipe) {
            addLeatherRecipe();
        }

        if (ConfigHandler.enableStringRecipe) {
            addStringRecipe();
        }

        if (ConfigHandler.enableWebRecipe) {
            addWebRecipe();
        }

        if (ConfigHandler.enableGlowstoneRecipe) {
            addGlowstoneRecipe();
        }

        if (ConfigHandler.enableGunpowderRecipe) {
            addGunpowderRecipe();
        }

        if (ConfigHandler.enableNetherWartRecipe) {
            addNetherWartRecipe();
        }

        if (ConfigHandler.enableNetherQuartzRecipe) {
            addNetherQuartzRecipe();
        }

        if (ConfigHandler.enablePrismarineCrystalRecipe) {
            addPrismarineCrystalRecipe();
        }

        if (ConfigHandler.enablePrismarineShardRecipe) {
            addPrismarineShardRecipe();
        }
    }

    private static void addSlimeBallRecipe() {
        ItemStack clayBall = new ItemStack(Items.CLAY_BALL);
        ItemStack sugar = new ItemStack(Items.SUGAR);
        ItemStack greenDye = new ItemStack(Items.DYE, 1, 2);    // Cactus green

        GameRegistry.addShapelessRecipe(new ItemStack(Items.SLIME_BALL), clayBall, sugar, greenDye);
    }

    private static void addCharcoalRecipe() {
        GameRegistry.addSmelting(Blocks.SAPLING, new ItemStack(Items.COAL, 1, 1), 0.35f);
    }

    private static void addFlintRecipe() {
        GameRegistry.addSmelting(Blocks.GRAVEL, new ItemStack(Items.FLINT), 0.35f);
    }

    private static void addLeatherRecipe() {
        GameRegistry.addSmelting(Items.ROTTEN_FLESH, new ItemStack(Items.LEATHER), 0.35f);
    }

    private static void addStringRecipe() {
        ItemStack web = new ItemStack(Blocks.WEB);

        GameRegistry.addShapelessRecipe(new ItemStack(Items.STRING), web);
    }

    private static void addWebRecipe() {
        GameRegistry.addRecipe(new ItemStack(Blocks.WEB), new Object[] {
                "S.S",
                ".S.",
                "S.S",
                'S', Items.STRING
        });
    }

    private static void addGlowstoneRecipe() {
        GameRegistry.addRecipe(new ItemStack(Items.GLOWSTONE_DUST), new Object[] {
                "NYN",
                "YRY",
                "NYN",
                'N', Items.GOLD_NUGGET,
                'Y', new ItemStack(Items.DYE, 1 , 11),  // Dandelion yellow
                'R', Items.REDSTONE
        });
    }

    private static void addGunpowderRecipe() {
        GameRegistry.addRecipe(new ItemStack(Items.GUNPOWDER), new Object[] {
                "CHC",
                "SFS",
                "CHC",
                'C', Items.CLAY_BALL,
                'H', new ItemStack(Items.COAL, 1, 1),   // Charcoal
                'S', Items.SUGAR,
                'F', Items.FLINT
        });
    }

    private static void addNetherWartRecipe() {
        GameRegistry.addRecipe(new ItemStack(Items.NETHER_WART), new Object[] {
                ".B.",
                "FDF",
                ".B.",
                'B', new ItemStack(Items.DYE, 1, 15),   // Bone meal
                'F', Items.FERMENTED_SPIDER_EYE,
                'D', new ItemStack(Blocks.DEADBUSH)
        });
    }

    private static void addNetherQuartzRecipe() {
        GameRegistry.addRecipe(new ItemStack(Items.QUARTZ), new Object[] {
                "DAD",
                "BSB",
                "DAD",
                'A', new ItemStack(Blocks.STONE, 1, 5), // Andesite
                'D', new ItemStack(Blocks.STONE,1, 3),  // Diorite
                'B', new ItemStack(Items.DYE, 1, 15),   // Bone meal
                'S', Items.SUGAR

        });
    }

    private static void addPrismarineCrystalRecipe() {
        GameRegistry.addRecipe(new ItemStack(Items.PRISMARINE_CRYSTALS, 2), new Object[] {
                "LSL",
                "RGR",
                "LSL",
                'L', new ItemStack(Items.DYE, 1, 4),    // Lapis lazuli
                'S', Items.SUGAR,
                'R', Items.REDSTONE,
                'G', Items.GLOWSTONE_DUST
        });
    }

    private static void addPrismarineShardRecipe() {
        GameRegistry.addRecipe(new ItemStack(Items.PRISMARINE_SHARD, 2), new Object[] {
                "CLC",
                "QAQ",
                "CLC",
                'C', new ItemStack(Blocks.COBBLESTONE),
                'L', new ItemStack(Items.DYE, 1, 4),    // Lapis lazuli
                'Q', Items.QUARTZ,
                'A', new ItemStack(Blocks.STONE, 1, 5), // Andesite
        });
    }
}


src/main/java/com/icaroperseo/helpfultweaks/client/core/config/ConfigHandler.java
package com.icaroperseo.helpfultweaks.client.core.config;

/**
 * Created by icaroperseo on 1/03/17.
 */

import net.minecraftforge.fml.common.Loader;
import net.minecraftforge.common.config.Configuration;
import net.minecraftforge.common.config.Property;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

public class ConfigHandler {
    public static boolean enableCharcoalRecipe;
    public static boolean enableFlintRecipe;
    public static boolean enableSlimeBallRecipe;
    public static boolean enableLeatherRecipe;
    public static boolean enableStringRecipe;
    public static boolean enableWebRecipe;
    public static boolean enableGlowstoneRecipe;
    public static boolean enableGunpowderRecipe;
    public static boolean enableNetherWartRecipe;
    public static boolean enableNetherQuartzRecipe;
    public static boolean enablePrismarineCrystalRecipe;
    public static boolean enablePrismarineShardRecipe;

    public static final String CATEGORY_NAME_RECIPES = "Recipes";
    public static final String CATEGORY_NAME_SMELTING = "Smelting";

    public static void preInit() {
        File configFile = new File(Loader.instance().getConfigDir(), "helpfultweaks.cfg");

        config = new Configuration(configFile);

        loadConfigFromFile();
    }

    private static void loadConfigFromFile() {
        /**
         * Load raw values from config file.
         */
        config.load();

        /**
         * Define the properties in the configuration file.
         */
        final boolean ENABLE_SLIME_BALL_RECIPE_DEFAULT_VALUE = true;
        Property propEnableSlimeBallRecipe = config.get(CATEGORY_NAME_RECIPES, "enableSlimeBallRecipe",
                ENABLE_SLIME_BALL_RECIPE_DEFAULT_VALUE, "Enables crafting slime balls recipe.");

        final boolean ENABLE_STRING_RECIPE_DEFAULT_VALUE = true;
        Property propEnableStringRecipe = config.get(CATEGORY_NAME_RECIPES, "enableStringRecipe",
                ENABLE_STRING_RECIPE_DEFAULT_VALUE, "Enables crafting web to string.");

        final boolean ENABLE_WEB_RECIPE_DEFAULT_VALUE = true;
        Property propEnableWebRecipe = config.get(CATEGORY_NAME_RECIPES, "enableWebRecipe",
                ENABLE_WEB_RECIPE_DEFAULT_VALUE, "Enables crafting string to web.");

        final boolean ENABLE_CHARCOAL_RECIPE_DEFAULT_VALUE = true;
        Property propEnableCharcoalRecipe = config.get(CATEGORY_NAME_SMELTING, "enableCharcoalRecipe",
                ENABLE_CHARCOAL_RECIPE_DEFAULT_VALUE, "Enables smelting saplings to charcoal.");

        final boolean ENABLE_FLINT_RECIPE_DEFAULT_VALUE = true;
        Property propEnableFlintRecipe = config.get(CATEGORY_NAME_SMELTING, "enableFlintRecipe",
                ENABLE_FLINT_RECIPE_DEFAULT_VALUE, "Enables smelting gravel to flint.");

        final boolean ENABLE_LEATHER_RECIPE_DEFAULT_VALUE = true;
        Property propEnableLeatherRecipe = config.get(CATEGORY_NAME_SMELTING, "enableLeatherRecipe",
                ENABLE_LEATHER_RECIPE_DEFAULT_VALUE, "Enables smelting rotten flesh to leather.");

        final boolean ENABLE_GLOWSTONE_RECIPE_DEFAULT_VALUE = true;
        Property propEnableGlowstoneRecipe = config.get(CATEGORY_NAME_RECIPES, "enableGlowstoneRecipe",
                ENABLE_GLOWSTONE_RECIPE_DEFAULT_VALUE, "Enables glowstone recipe.");

        final boolean ENABLE_GUNPOWDER_RECIPE_DEFAULT_VALUE = true;
        Property propEnableGunpowderRecipe = config.get(CATEGORY_NAME_RECIPES, "enableGunpowderRecipe",
                ENABLE_GUNPOWDER_RECIPE_DEFAULT_VALUE, "Enables gunpowder recipe.");

        final boolean ENABLE_NETHER_WART_RECIPE_DEFAULT_VALUE = true;
        Property propEnableNetherWartRecipe = config.get(CATEGORY_NAME_RECIPES, "enableNetherWartRecipe",
                ENABLE_NETHER_WART_RECIPE_DEFAULT_VALUE, "Enables nether wart recipe.");

        final boolean ENABLE_NETHER_QUARTZ_RECIPE_DEFAULT_VALUE = true;
        Property propEnableNetherQuartzRecipe = config.get(CATEGORY_NAME_RECIPES, "enableNetherQuartzRecipe",
                ENABLE_NETHER_QUARTZ_RECIPE_DEFAULT_VALUE, "Enables nether quartz recipe.");

        final boolean ENABLE_PRISMARINE_CRYSTAL_RECIPE_DEFAULT_VALUE = true;
        Property propEnablePrismarineCrystalRecipe = config.get(CATEGORY_NAME_RECIPES, "enablePrismarineCrystalRecipe",
                ENABLE_PRISMARINE_CRYSTAL_RECIPE_DEFAULT_VALUE, "Enables prismarine crystal recipe.");

        final boolean ENABLE_PRISMARINE_SHARD_RECIPE_DEFAULT_VALUE = true;
        Property propEnablePrismarineShardRecipe = config.get(CATEGORY_NAME_RECIPES, "enablePrismarineShardRecipe",
                ENABLE_PRISMARINE_SHARD_RECIPE_DEFAULT_VALUE, "Enables prismarine shard recipe.");

        List<String> propOrderRecipes = new ArrayList<String>();

        propOrderRecipes.add(propEnableSlimeBallRecipe.getName());
        propOrderRecipes.add(propEnableStringRecipe.getName());
        propOrderRecipes.add(propEnableWebRecipe.getName());
        propOrderRecipes.add(propEnableGlowstoneRecipe.getName());
        propOrderRecipes.add(propEnableGunpowderRecipe.getName());
        propOrderRecipes.add(propEnableNetherWartRecipe.getName());
        propOrderRecipes.add(propEnableNetherQuartzRecipe.getName());
        propOrderRecipes.add(propEnablePrismarineCrystalRecipe.getName());
        propOrderRecipes.add(propEnablePrismarineShardRecipe.getName());
        config.setCategoryPropertyOrder(CATEGORY_NAME_RECIPES, propOrderRecipes);

        List<String> propOrderSmelting = new ArrayList<String>();

        propOrderSmelting.add(propEnableCharcoalRecipe.getName());
        propOrderSmelting.add(propEnableFlintRecipe.getName());
        propOrderSmelting.add(propEnableLeatherRecipe.getName());
        config.setCategoryPropertyOrder(CATEGORY_NAME_SMELTING, propOrderSmelting);

        /**
         * Read the configuration property values into the class's variables.
         */
        enableSlimeBallRecipe = propEnableSlimeBallRecipe.getBoolean(ENABLE_SLIME_BALL_RECIPE_DEFAULT_VALUE);
        enableStringRecipe = propEnableStringRecipe.getBoolean(ENABLE_STRING_RECIPE_DEFAULT_VALUE);
        enableCharcoalRecipe = propEnableCharcoalRecipe.getBoolean(ENABLE_CHARCOAL_RECIPE_DEFAULT_VALUE);
        enableFlintRecipe = propEnableFlintRecipe.getBoolean(ENABLE_FLINT_RECIPE_DEFAULT_VALUE);
        enableLeatherRecipe = propEnableLeatherRecipe.getBoolean(ENABLE_LEATHER_RECIPE_DEFAULT_VALUE);
        enableWebRecipe = propEnableWebRecipe.getBoolean(ENABLE_WEB_RECIPE_DEFAULT_VALUE);
        enableGlowstoneRecipe = propEnableGlowstoneRecipe.getBoolean(ENABLE_GLOWSTONE_RECIPE_DEFAULT_VALUE);
        enableGunpowderRecipe = propEnableGunpowderRecipe.getBoolean(ENABLE_GUNPOWDER_RECIPE_DEFAULT_VALUE);
        enableNetherWartRecipe = propEnableNetherWartRecipe.getBoolean(ENABLE_NETHER_WART_RECIPE_DEFAULT_VALUE);
        enableNetherQuartzRecipe = propEnableNetherQuartzRecipe.getBoolean(ENABLE_NETHER_QUARTZ_RECIPE_DEFAULT_VALUE);
        enablePrismarineCrystalRecipe =
                propEnablePrismarineCrystalRecipe.getBoolean(ENABLE_PRISMARINE_CRYSTAL_RECIPE_DEFAULT_VALUE);
        enablePrismarineShardRecipe =
                propEnablePrismarineShardRecipe.getBoolean(ENABLE_PRISMARINE_SHARD_RECIPE_DEFAULT_VALUE);

        /**
         * Write the class's variables back into the config properties and save to disk.
         *
         * This is done because some of the properties may have been assigned default values if the file was empty or
         * corrupt.
         */
        propEnableSlimeBallRecipe.set(enableSlimeBallRecipe);
        propEnableStringRecipe.set(enableStringRecipe);
        propEnableCharcoalRecipe.set(enableCharcoalRecipe);
        propEnableFlintRecipe.set(enableFlintRecipe);
        propEnableLeatherRecipe.set(enableLeatherRecipe);
        propEnableWebRecipe.set(enableWebRecipe);
        propEnableGlowstoneRecipe.set(enableGlowstoneRecipe);
        propEnableGunpowderRecipe.set(enableGunpowderRecipe);
        propEnableNetherWartRecipe.set(enableNetherWartRecipe);
        propEnableNetherQuartzRecipe.set(enableNetherQuartzRecipe);
        propEnablePrismarineCrystalRecipe.set(enablePrismarineCrystalRecipe);
        propEnablePrismarineShardRecipe.set(enablePrismarineShardRecipe);

        if (config.hasChanged()) {
            config.save();
        }
    }

    private static Configuration config = null;
}
